src/components/ui/Suggest.jsx

```diff
@@ -0,0 +1,134 @@
+  let currentQuery = null;
+
+  useEffect(() => {
+    const handleFocus = () => setIsOpen(true);
+
+    const handleBlur = () => {
+      setIsHighlighted(false);
+      setIsOpen(false);
+    };
+
+    const handleInput = debounce(e => {
+      const typedValue = e.target.value;
+      setIsPending('fetching');
+
+      if (currentQuery) {
+        currentQuery.abort();
+      }
+
+      const query = suggestResults(typedValue, {
+        withCategories,
+        useFocus: SUGGEST_USE_FOCUS,
+        focusMinZoom: SUGGEST_FOCUS_MIN_ZOOM,
+        maxFavorites: !typedValue ? 5 : 2,
+        maxItems: SUGGEST_MAX_ITEMS,
+      });
+
+      currentQuery = query;
+      setLastQuery(typedValue);
+
+      query
+        .then(suggestions => modifyList(suggestions, withGeoloc))
+        .then(items => {
+          setItems(items);
+          setIsPending('completed');
+          currentQuery = null;
+        })
+        .catch(() => { /* Query aborted. Just ignore silently */ });
+    }, 100);
+
+    searchInputDomHandler.addEventListener('focus', handleFocus);
+    searchInputDomHandler.addEventListener('blur', handleBlur);
+    searchInputDomHandler.addEventListener('input', handleInput);
+
+    return () => {
+      searchInputDomHandler.removeEventListener('blur', handleFocus);
+      searchInputDomHandler.removeEventListener('blur', handleBlur);
+      searchInputDomHandler.removeEventListener('input', handleInput);
+    };
+  }, []);
+
+  useEffect(() => {
+    const handleSubmit = () => {
+      selectItem(items[0], true);
+    };
+
+    const handleKeyDown = async event => {
+      if (event.key === 'Esc' || event.key === 'Escape') { // esc
+        setIsOpen(false);
+        setIsHighlighted(false);
+      } else if (event.key === 'Enter' && !isHighlighted) {
+        setIsPendingSubmit(true);
+        setIsOpen(false);
+      } else {
+        setIsOpen(true);
+      }
+    };
+
+    searchInputDomHandler.addEventListener('keydown', handleKeyDown);
+    searchInputDomHandler.addEventListener('submit', handleSubmit);
+
+    return () => {
+      searchInputDomHandler.removeEventListener('keydown', handleKeyDown);
+      searchInputDomHandler.removeEventListener('submit', handleSubmit);
+    };
+  }, [items, isHighlighted]);
+
+  useEffect(() => {
+    if (isPending === 'completed' && isPendingSubmit) {
+      selectItem(items[0], true);
+      setIsOpen(false);
+      setIsPendingSubmit(false);
+      setIsPending('unset');
+    }
+  }, [isPending, isPendingSubmit]);
+
+  if (!isOpen) {
+    return null;
+  }
+
+  return (
+    <SuggestsDropdown
+      inputId="search"
+      suggestItems={items}
+      onHighlight={item => {
+        if (!item) {
+          setIsHighlighted(false);
+          searchInputDomHandler.value = lastQuery;
+        } else {
+          setIsHighlighted(true);
+          searchInputDomHandler.value = item.name;
+        }
+      }}
+      onSelect={item => {
```

Comment:
You do a lot of things here that are done almost identically elsewhere. What do you think about defining a function like `onItemSelected(item)` and try to reuse it instead of doing everything here anonymously.

Reply:
Well, I would like but there are only two "selection" cases: With "onSelect" called by `SuggestDropdown` component, or a "selection" using the `Enter` key. Enter key is a special case because we have to wait until the query is resolved. I don't find a proper way to merge these two cases.